### ISSUE TITLE
Cron task executing only for 10 minutes when scheduled but not when "run now"

### ISSUE THREAD

**Issue by lbetineli** [AOEpeople/Aoe_Scheduler#329]
I'm using Magento 1.9.3.2 and what I face is:

I've developed a module with some crons... One of them call a lot of thirdy-part web services via soap and sometimes it simply throw that famous error "Error fetching http headers".
Another Cron has less soap calls and after 10 minutes it just stops working, look at the first and last log for this cron:

```
2017-04-04T 12:40:07+00:00 DEBUG (7): Updating product: 0001
2017-04-04T 12:50:00+00:00 DEBUG (7): Updating product: 00XX

```
After this I created a cron to run a little code like this one:

```
for($i=0; $i<20; $i++)
{
   Mage::log("Minute: " . $i, null, "Test.log");
   sleep(60);
}
```
And in all tests the code simply stopped in the 10th minute! (When I upload this same code above as a single php file in the server and run it from browser it completed 20 minutes normally)

I'm using AOE Scheduler and I've already changed some config, deactivated the module to use Magento cron manager, but nothing changed...

I also tried to set the "default_socket_time" in my .htaccess ('cause I don't have access to my php.ini on the server) but nothing...

Could it be some Magento/Module or server configuration? Anyone have already faced it?

I don't know what to try anymore, so if someone could help I'd appreciate :)

**Comment by LeeSaferite** [comment]
http://php.net/manual/en/function.set-time-limit.php

**Comment by LeeSaferite** [comment]
More than likely you have a cli config for PHP that sets the max execution time to 10 minutes.
http://php.net/manual/en/info.configuration.php#ini.max-execution-time

You could also look into dynamically extending the execution time limit periodically if you have an execution loop where it makes sense. http://php.net/manual/en/function.set-time-limit.php

**Comment by lbetineli** [comment]
But my "max_execution_time" is set in .htaccess to 18000 :( 

The fact that when I run this specific cron using the Run Now is that the context is other than CLI? (just to understand why when I run using it the whole script is run)

**Comment by LeeSaferite** [comment]
The PHP settings in the .htaccess file only apply to web requests. The cron process runs as a CLI request and would use the PHP ini file. Depending on your server it may share the PHP ini between web and CLI, but some setups have a different ini file for CLI requests. When you run a cron job using 'Run Now' you are actually executing it differently than a normal cron job. It's executed in the web request context and using the web server user. This presents a few issues and is why it's disabled by default. If your cron jobs create files, then those files will belong to the web user and not the cron user. They may be the same on your server, but are not required to be the same and may be different for security reasons.

To recap:
The settings in .htaccess are not used when the cron jobs are run from the standard cron process.

**Comment by lbetineli** [comment]
My phpinfo() gives me this:

max_execution_time	18000	3000
max_file_uploads	20	20
max_input_nesting_level	64	64
max_input_time	18000	18000
max_input_vars	100000	100000
memory_limit	1024M	1024M

In which the first columns is the "local" value and the second column the "master" value.

I'm running the cron mentioned above and it only run for 10 minutes... but if the max_execution_time is 3000 it doesn't make sense to make my cron disappear after 10 minutes... I tried now to put `$setTime = set_time_limit(0);` before the method starts but the same 10 minutes were executed... I really don't know what to do anymore.. :(

**Comment by lbetineli** [comment]
Now I think it is related to my php server api, which is CGI/FAST-CGI and I think my cron must be executed using the CLI interface, that has no time limit for executing crons, right?

I saw this http://www.htmlcenter.com/blog/running-php-scripts-with-cron/ and then tried to put the `#!/usr/local/bin/php -q` above the shell/scheduler.php file but nothing changed... any ideas?

(to find my php path I executed "whereis php" and the return was php: /usr/bin/php /usr/lib/php.ini /usr/lib/php /usr/local/bin/php /usr/local/lib/php.ini /usr/local/lib/php.ini,v /usr/local/lib/php /usr/include/php /usr/local/php)

**Comment by hostep** [comment]
@lbetineli: have you tried the `Schedule Now` option instead of the `Run Now` option?

The second one will always run in the context of the webserver, which is bad for long running tasks.
The first one will just schedule the job to be run the next time Magento's cronjob is triggered. If you have setup the cronjob [correctly](https://github.com/AOEpeople/Aoe_Scheduler/blob/master/doc/instructions.md), it should then run in the cli context.

Normally the cli context is configured to have no time limits, but it might be different on your particular hosting, you can check it like this (`0` is unlimited):
```bash
$ php -i | grep max_execution_time
max_execution_time => 0 => 0
```

Also make sure your code isn't calling the [`set_time_limit`](http://php.net/manual/en/function.set-time-limit.php) function anywhere, as that might influence the limit as well.

Hope this helps :)

**Comment by lbetineli** [comment]
Hi @hostep , thank you but it stills running 10 minutes... 

I found that if I run the Magento cron.sh via SSH (real php-cli) the code runs only 15 minutes also... I tried everything including setting time limit to 0, adding `#!/usr/bin/php-cli -q`, but nothing seems to make difference...

I am in contact with my hosting company to see if they help but nothing till now :(

Is is wrong to have a cron that has to run for over an hour? I think it is not, right?

**Comment by hostep** [comment]
No it's not wrong to have long running tasks, if it is running using the cli it shouldn't time out.

I'm sure your hosting company will be able to help you further, that's usually the best thing to do is to ask the guys who setup your server, as they should know these kind of things :)

**Comment by lbetineli** [comment]
I found via SSH that there is a PHP-CLI on the server and then my cron in cPanel became this `/usr/bin/php-cli -q /home/pbstorec/public_html/cron.php` but even this way my `echo PHP_SAPI` was `cgi-fcgi`.. it doesn't make sense at all...

My hosting said that even I putting "php-cli" in the cPanel the cPanel itself changes it for the normal php, which is the 'cgi-fcgi'... but I still think it is strange..

**Comment by lbetineli** [comment]
Which is right? Call the "Cron.sh" or the "Cron.php"? (Considering my test right now that I deactivated the AOE Scheduler)

**Comment by LeeSaferite** [comment]
Always the bash shell script. It has some checks to prevent parallel execution of the cron system.

**Comment by lbetineli** [comment]
I couldn't really find out what was going on... what I did was to configure a separate cron file to my method and configure it directly on the cpanel. Funny fact: when I ran a 30-minute test method directly on the cPanel cron it was ok, but if I put the same method to run inside the Magento cron.php it only ran 10 minutes... It seemed that Magento was limiting the execution... but thank all of you for the help!
